onnxruntime/core/providers/cuda/tensor/transpose_impl.cu

```diff
@@ -104,63 +105,124 @@ __global__ void Transpose4DKernel(const TArray<int64_t> input_strides, const voi
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Type not supported for transpose on CUDA. Element size was ",
+                           element_size);
+  }
+
   unsigned int num_elements_per_thread = 4 * sizeof(int) / static_cast<unsigned int>(element_size);  // int4 is used in the kernel to access data.
   dim3 block_size(static_cast<unsigned int>(input_shape[3] / num_elements_per_thread), static_cast<unsigned int>(input_shape[2]));
   dim3 grid_size(static_cast<unsigned int>(input_shape[1]), static_cast<unsigned int>(input_shape[0]));
 
-  switch (element_size) {
-    case sizeof(int8_t):
-      Transpose4DKernel<sizeof(int8_t)><<<grid_size, block_size, 0, stream>>>(
-          input_strides, input_data,
-          output_strides, output_data, N / num_elements_per_thread);
-      break;
-    case sizeof(int16_t):
-      Transpose4DKernel<sizeof(int16_t)><<<grid_size, block_size, 0, stream>>>(
-          input_strides, input_data,
-          output_strides, output_data, N / num_elements_per_thread);
-      break;
-    case sizeof(int32_t):
-      Transpose4DKernel<sizeof(int32_t)><<<grid_size, block_size, 0, stream>>>(
-          input_strides, input_data,
-          output_strides, output_data, N / num_elements_per_thread);
-      break;
-    case sizeof(int64_t):
-      Transpose4DKernel<sizeof(int64_t)><<<grid_size, block_size, 0, stream>>>(
-          input_strides, input_data,
-          output_strides, output_data, N / num_elements_per_thread);
-      break;
-    default:
-      return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Type not supported for transpose on CUDA. Element size was ",
-                             element_size);
+  Transpose4DKernelParallelizeMultipleElementsPerThreadInInnermostDim<<<grid_size, block_size, 0, stream>>>(
+      input_strides, input_data,
+      output_strides, output_data, element_size, N / num_elements_per_thread);
+
+  return Status::OK();
+}
+
+__global__ void Transpose4DKernelParallelizeOneElementPerThread(
+    const TArray<int64_t> input_strides, const int8_t* input_data,
+    const TArray<int64_t> output_strides, int8_t* output_data,
+    size_t element_size,
+    CUDA_LONG N) {
+  CUDA_LONG input_index = blockIdx.y * input_strides[0] +
+                          blockIdx.x * input_strides[1] +
+                          threadIdx.y * input_strides[2] +
+                          threadIdx.x * input_strides[3];
+
+  CUDA_LONG output_index = blockIdx.y * output_strides[0] +
+                           blockIdx.x * output_strides[1] +
+                           threadIdx.y * output_strides[2] +
+                           threadIdx.x * output_strides[3];
+
+  if (input_index < N && output_index < N) {
+    const int8_t* input_data_to_be_copied = input_data + (input_index * element_size);
+    int8_t* output_data_to_be_copied = output_data + (output_index * element_size);
+
+    // copy over the bytes
+    memcpy(output_data_to_be_copied, input_data_to_be_copied, element_size);
```

Comment:
According to this link, https://stackoverflow.com/questions/10456728/is-there-an-equivalent-to-memcpy-that-works-inside-a-cuda-kernel/10468720, memcpy may not be optimal. Can you write some tests with different tensor size and measure the kernel execution's time?

Reply:
We an skip the system calls from the kernel - I think this is a good observation. We can simply loop over the bytes and copy it over. I can measure the latency but my guess is without the memcpy and using the looping logic, it should still perform better than the generic implementation.